src/GazeboUnderwater.cpp

```diff
@@ -20,82 +20,140 @@ namespace gazebo_underwater
+                physics::InertialPtr modelInertia = link->GetInertial();
+                gzmsg << "GazeboUnderwater: link mass: " << modelInertia->GetMass() << std::endl;
+            }else{
+                std::string msg = "GazeboUnderwater: link " + sdf->Get<std::string>("link_name")
+                        + " not found in model " + model->GetName();
+                gzthrow(msg);
+            }
+        }else{
+            gzthrow("GazeboUnderwater: link_name not defined in world file !");
+        }
+
+        waterLevel = getParameter<double>("water_level","meters", 0.0);
+        fluidVelocity = getParameter<math::Vector3>("fluid_velocity","m/s",math::Vector3(0,0,0));
+        densityOfFluid = getParameter<double>("fluid_density","kg/m3", 1027);
+        dragCoefficient = getParameter<math::Vector3>("drag_coefficient","dimensionless",
+                math::Vector3(1,1,1));
+        // buoyancy must be the buoyancy when the model is completely submersed
+        buoyancy = getParameter<double>("buoyancy","N",0);
         centerOfBuoyancy = getParameter<math::Vector3>("center_of_buoyancy","meters",
-                math::Vector3(0, 0, 0.1));
-        fluidVelocity = getParameter<math::Vector3>("fluid_velocity","m/s",
-                math::Vector3(0,0,0));
-        viscousDamping = getParameter<math::Vector3>("viscous_damping","dimensionless",
-                math::Vector3(20,30,50));
-        waterLevel = getParameter<double>("water_level","meters", 2.0);
-        densityOfFluid = getParameter<double>("fluid_density","kg/m3", 1000);
+                math::Vector3(0, 0, 0.15));
+        sideAreas = getParameter<math::Vector3>("side_areas","meter2",
+                math::Vector3(0.5,0.5,0.5));
+        volume = getParameter<double>("volume","meter3",1);
     }
 
     void GazeboUnderwater::updateBegin(common::UpdateInfo const& info)
     {
-        physics::LinkPtr link = model->GetLink( sdf->Get<std::string>("link_name") ); 
-        applyBuoyancy(link);
-        applyViscousFriction(link);
+        applyBuoyancy();
+        applyViscousDrag();
     }
 
-    void GazeboUnderwater::applyViscousFriction(physics::LinkPtr link)
+    void GazeboUnderwater::applyViscousDrag()
     {
-        // Calculates dynamic viscous damp
-        math::Vector3 fluidVelocityRelative =
-            link->GetWorldPose().rot.RotateVectorReverse(fluidVelocity);
-        math::Vector3 velocityDifference =
-            link->GetRelativeLinearVel() - fluidVelocityRelative;
-        link->AddRelativeForce(- velocityDifference * viscousDamping);
+        math::Vector3 cogPosition = link->GetWorldCoGPose().pos;
+        double distanceToSurface = waterLevel - cogPosition.z;
+        if(distanceToSurface > 0 )
+        {
+            math::Vector3 velocityDifference = link->GetWorldCoGLinearVel() - fluidVelocity;
+            math::Vector3 viscousDrag = - 0.5 * densityOfFluid * sideAreas * dragCoefficient *
```

Comment:
Don't you need to project (ortho) the bounding box w.r.t the direction of the flow ?